src/kernel/pmm.zig

```diff
@@ -0,0 +1,214 @@
+    NotAllocated,
+};
+
+/// The size of memory associated with each bitmap entry
+const BLOCK_SIZE = arch.MEMORY_BLOCK_SIZE;
+
+var bitmap: PmmBitmap = undefined;
+
+///
+/// Set the bitmap entry for an address as occupied
+///
+/// Arguments:
+///     IN addr: usize - The address.
+///
+/// Error: PmmBitmap.BitmapError.
+///     *: See PmmBitmap.setEntry. Could occur if the address is out of bounds.
+///
+fn setAddr(addr: usize) PmmBitmap.BitmapError!void {
+    try bitmap.setEntry(@intCast(u32, addr / BLOCK_SIZE));
+}
+
+///
+/// Check if an address is set as occupied.
+///
+/// Arguments:
+///     IN addr: usize - The address to check.
+///
+/// Return: True if occupied, else false.
+///
+/// Error: PmmBitmap.BitmapError.
+///     *: See PmmBitmap.setEntry. Could occur if the address is out of bounds.
+///
+fn isSet(addr: usize) PmmBitmap.BitmapError!bool {
+    return bitmap.isSet(@intCast(u32, addr / BLOCK_SIZE));
+}
+
+///
+/// Find the next free memory block, set it as occupied and return it. The region allocated will be of size BLOCK_SIZE.
+///
+/// Return: The address that was allocated.
+///
+pub fn alloc() ?usize {
+    if (bitmap.setFirstFree()) |entry| {
+        return entry * BLOCK_SIZE;
+    }
+    return null;
+}
+
+///
+/// Set the address as free so it can be allocated in the future. This will free a block of size BLOCK_SIZE.
+///
+/// Arguments:
+///     IN addr: usize - The previously allocated address to free. Will be aligned down to the nearest multiple of BLOCK_SIZE.
+///
+/// Error: Error || PmmBitmap.BitmapError.
+///     Error.NotAllocated: The address wasn't allocated.
+///     PmmBitmap.BitmapError.OutOfBounds: The address given was out of bounds.
+///
+pub fn free(addr: usize) (PmmBitmap.BitmapError || Error)!void {
+    const idx = @intCast(u32, addr / BLOCK_SIZE);
+    if (try bitmap.isSet(idx)) {
+        try bitmap.clearEntry(idx);
+    } else {
+        return Error.NotAllocated;
+    }
+}
+
+///
+/// Intiialise the physical memory manager and set all unavailable regions as occupied (those from the memory map and those from the linker symbols).
+///
+/// Arguments:
+///     IN mem: *const MemProfile - The system's memory profile.
+///     IN allocator: *std.mem.Allocator - The allocator to use to allocate the bitmaps.
+///
+pub fn init(mem: *const MemProfile, allocator: *std.mem.Allocator) void {
+    log.logInfo("Init pmm\n", .{});
+    bitmap = PmmBitmap.init(mem.mem_kb * 1024 / BLOCK_SIZE, allocator) catch @panic("Bitmap allocation failed");
+
+    // Occupy the regions of memory that the memory map describes as reserved
+    for (mem.mem_map) |entry| {
+        if (entry.@"type" != MEMORY_AVAILABLE) {
+            var addr = std.mem.alignBackward(@intCast(usize, entry.addr), BLOCK_SIZE);
+            var end = @intCast(usize, entry.addr + (entry.len - 1));
+            // If the end address can be aligned without overflowing then align it
+            if (end <= std.math.maxInt(usize) - BLOCK_SIZE)
```

Comment:
Idk whether we should have `{}` after the if statements. Or at least have a space before the while loop so is clearer